### PR TITLE
net: lib: lwm2m: add lpwan object

### PR DESCRIPTION
**Pull request by marco85mars20**
Add support for Low Power Wide Area Network (LPWAN) Object (ID 3412)

Signed-off-by: Marco Argiolas <marco.argiolas@ftpsolutions.com.au>

### CODE REVIEW

**Review by rlubos — COMMENTED** [review]
Looks pretty good, just some remarks regarding naming.

File: subsys/net/lib/lwm2m/Kconfig.ucifi [hunk 1]
```diff
@@ -25,4 +25,87 @@ config LWM2M_UCIFI_BATTERY_INSTANCE_COUNT
 
 endif # LWM2M_UCIFI_BATTERY
 
+menuconfig LWM2M_UCIFI_LPWAN
+	bool "uCIFI LPWAN (3412) support"
+	help
+	  Include support for LWM2M Connectivity Monitoring Object (ID 4)
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m/Kconfig.ucifi:31]: Mismatch of help text and object name. This is not a Connectivity Monitoring object.

Comment (marco85mars20) [reply subsys/net/lib/lwm2m/Kconfig.ucifi]: It has now been fixed.
Thanks

**Comment by marco85mars20** [comment]
@rlubos 
I believe I have applied all your suggested changes. Thanks for your review

File: subsys/net/lib/lwm2m/Kconfig.ucifi [hunk 2]
```diff
@@ -25,4 +25,88 @@ config LWM2M_UCIFI_BATTERY_INSTANCE_COUNT
...
+	bool "uCIFI LPWAN (3412) Support"
+	help
+	  The uCIFI LPWAN object provides attributes to monitor and maintain
+	  Low-Power Wide Area Networks.
+
+if LWM2M_UCIFI_LPWAN
+
+config LWM2M_UCIFI_LPWAN_INSTANCE_COUNT
+	int "Maximum # of uCIFI LPWAN object instances"
+	depends on LWM2M_UCIFI_LPWAN
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m/Kconfig.ucifi:38]:
```suggestion
```

All these configuration parameters are already under `if LWM2M_UCIFI_LPWAN` so they don't need separate "depends on` clause.

**Review by SeppoTakalo — CHANGES_REQUESTED** [review]
Looks good.

Just drop those extra `depends on` clauses.

File: subsys/net/lib/lwm2m/Kconfig.ucifi (hunk 2 above)
Comment (marco85mars20) [reply subsys/net/lib/lwm2m/Kconfig.ucifi]: This has been fixed too.
Thanks

**Review by rlubos — COMMENTED** [review]
Thanks @marco85mars20, look good to me.

One final request from my side - please squash the two commits, there's no need to add files in one commit only to remove/rename them in the next one. We strive to keep the commit history clean, thanks!

**Comment by marco85mars20** [comment]
> Thanks @marco85mars20, look good to me.
> 
> One final request from my side - please squash the two commits, there's no need to add files in one commit only to remove/rename them in the next one. We strive to keep the commit history clean, thanks!

Sure, I will.
I thought PRs are squashed by github when merging.

File: subsys/net/lib/lwm2m/Kconfig.ucifi [hunk 3]
```diff
@@ -25,4 +25,80 @@ config LWM2M_UCIFI_BATTERY_INSTANCE_COUNT
...
+	help
+	  This value sets the maximum allowable number of LPWAN instances.
+
+config LWM2M_UCIFI_LPWAN_IPV4_ADDRESS_MAX
+	int "Maximum # of IPv4 addresses"
+	default 0
+	range 0 20
+	help
+	  This value sets the maximum number of IPv4 address assigned to a
+	  device that can be stored in this object.
```
Comment (nixward) [inline subsys/net/lib/lwm2m/Kconfig.ucifi:49]:
```suggestion
	  This value sets the maximum number of IPv4 addresses assigned to a
	  device that can be stored in this object.
```

File: subsys/net/lib/lwm2m/Kconfig.ucifi [hunk 4]
```diff
@@ -25,4 +25,80 @@ config LWM2M_UCIFI_BATTERY_INSTANCE_COUNT
...
+	int "Maximum # of secondary network addresses"
+	default 0
+	range 0 20
+	help
+	  This value sets the maximum number of secondary addresses used to
+	  communicate with the device on the LPWAN network assigned that can be
+	  stored in this object.
+
+config LWM2M_UCIFI_LPWAN_PEER_ADDRESS_MAX
+	int "Maximum # of addresses of peer nodes"
```
Comment (nixward) [inline subsys/net/lib/lwm2m/Kconfig.ucifi:77]:
```suggestion
	int "Maximum # of peer node addresses"
```

File: subsys/net/lib/lwm2m/Kconfig.ucifi [hunk 5]
```diff
@@ -25,4 +25,80 @@ config LWM2M_UCIFI_BATTERY_INSTANCE_COUNT
...
+	  that can be stored in this object.
+
+config LWM2M_UCIFI_LPWAN_MULTICAST_GRP_ADDRESS_MAX
+	int "Maximum # of multicast addresses"
+	default 0
+	range 0 20
+	help
+	  This value sets the maximum number of group addresses (from which the
+	  device should accept incoming messages and/or commands) that can be
+	  stored in this object.
```
Comment (nixward) [inline subsys/net/lib/lwm2m/Kconfig.ucifi:91]:
```suggestion
	int "Maximum # of multicast group addresses"
	default 0
	range 0 20
	help
	  This value sets the maximum number of multicast group addresses (from
	  which the device should accept incoming messages and/or commands)
	  that can be stored in this object.
```

File: subsys/net/lib/lwm2m/ucifi_lpwan.c [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+	}
+
+	if (index >= MAX_INSTANCE_COUNT) {
+		LOG_ERR("Can not create instance - no more room: %u", obj_inst_id);
+		return NULL;
+	}
+
+	/* Set default values */
+	(void)memset(res[index], 0, sizeof(res[index][0]) * ARRAY_SIZE(res[index]));
+	(void)memset(mac[index], 0, MAC_ADDRESS_LEN);
```
Comment (nixward) [inline subsys/net/lib/lwm2m/ucifi_lpwan.c:122]:
```suggestion
	(void)memset(res[index], 0, sizeof(res[index]));
	mac[index][0] = '\0';
```

**Comment by marco85mars20** [comment]
Thanks @nixward for your feedbacks, I have applied your suggestions.
I have also done the following:
- added `CONFIG_LWM2M_UCIFI_SUPPORT` in `tests/net/all/prj.conf`
- fixed a couple of bugs related to renamed macros

File: subsys/net/lib/lwm2m/ucifi_lpwan.c [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+/* resource state variables */
+static char mac[MAX_INSTANCE_COUNT][MAC_ADDRESS_SIZE];
+
+static struct lwm2m_engine_obj_inst *lpwan_create(uint16_t obj_inst_id)
+{
+	int index, i = 0, j = 0;
+
+	/* Check that there is no other instance with this ID */
+	for (index = 0; index < MAX_INSTANCE_COUNT; index++) {
+		if (inst[index].obj && inst[index].obj_inst_id == obj_inst_id) {
```
Comment (nixward) [inline subsys/net/lib/lwm2m/ucifi_lpwan.c:102]:
```suggestion
		if ((inst[index].obj != NULL) && inst[index].obj_inst_id == obj_inst_id) {
```

Comment (marco85mars20) [reply subsys/net/lib/lwm2m/ucifi_lpwan.c]: I have applied this change too.

**Review by rlubos — APPROVED** [review]

**Review by SeppoTakalo — APPROVED** [review]

**Pull request merged**